src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp

```diff
@@ -75,26 +75,34 @@ QVariant QgsValueRelationWidgetWrapper::value() const
       }
     }
 
-    QVariantList vl;
-    //store as QVariantList because it's json
-    for ( const QString &s : qgis::as_const( selection ) )
+    if ( layer()->fields().at( fieldIdx() ).type() == QVariant::Map )
     {
-      // Convert to proper type
-      const QVariant::Type type { fkType() };
-      switch ( type )
+      QVariantList vl;
+      //store as QVariantList because the field type supports data structure
+      for ( const QString &s : qgis::as_const( selection ) )
       {
-        case QVariant::Type::Int:
-          vl.push_back( s.toInt() );
-          break;
-        case QVariant::Type::LongLong:
-          vl.push_back( s.toLongLong() );
-          break;
-        default:
-          vl.push_back( s );
-          break;
+        // Convert to proper type
+        const QVariant::Type type { fkType() };
+        switch ( type )
+        {
+          case QVariant::Type::Int:
+            vl.push_back( s.toInt() );
+            break;
+          case QVariant::Type::LongLong:
+            vl.push_back( s.toLongLong() );
+            break;
+          default:
+            vl.push_back( s );
+            break;
+        }
       }
+      v = vl;
+    }
+    else
+    {
+      //store as a formatted string because the fields supports only string
+      v = selection.join( ',' ).prepend( '{' ).append( '}' );
```

Comment:
Are you sure about this?
Wouldn't be (much) better to use JSON in this case? Providers can convert from JSON to whatever internal format they use, depending on the field type, without loosing information. So, for instance, PG will convert to `hstore` only when needed, and keep JSON in case the field type accepts is.
OTOH, if you stick to `hstore` format, you risk any kind of issue with string keys containing `{},`

Reply:
I changed it back to how it has been before. 

I thought about to use JSON for that. But I wonder why it has been in this format at all. Is it just historical or is there a good reason for it? I don't know. Do you? Are there no risks using JSON format instead of HSTORE?

Reply:
I think that `hstore` was the first implementation of this feature, but it is native to PG.

In any event, `hstore` apparently support string keys, so if you want to use `hstore` safely you MUST change the logic that builds and `hstore` by properly quoting the keys (if they are not convertible to `int`), in other words: _never assume that keys are numeric_.

For instance (untested):
```
{ "a {} , \"key" => "a value \" with wierd chars {}," , 1234 => "A value with an int key" }

```

See: https://www.postgresql.org/docs/9.0/hstore.html

**The alternative would be to switch to JSON as the default representation.**

The only potential issue that comes up to my mind is that we might have existing legacy databases that rely on that `hstore` string representation of arrays, that might break if we switch to a JSON representation in case the same values are accessed by another application which is not QGIS (probably not that common use case though).

So, we might consider a transitory logic (until QGIS 4) that tries to guess if the storage format for a string field is `hstore` or `JSON` and act accordingly (issuing a warning in the logs it the format is `hstore`).

If the DB is empty we could just use JSON as a default string representation.

We shouldn't have a problem when reading values from the provider, because we still accept both representations (JSON and hstore).

**Conclusion**

I think that if the quoting of `hstore` keys works fine, we can continue to use `hstore` as a default string representation for arrays without risks. This has the advantage that we won't break any legacy code.

Switching to JSON has the advantage that we would be using a well known, well tested, standardized format with robust conversion libraries, but we would risk the legacy code issue mentioned early (even if I believe the risk is rather low).

Reply:
Thanks a lot.

So I see three possibilities:
- improve the `hstore` format (add quotes to text values) like you suggested above
- convert it in value relation widget (if no JSON-field) to a JSON formatted string
- keep it as a `QVariantList` (like before this pr) but add in OGR provider the functionality to convert (if no JSON-field) to a JSON formatted string (in case of the values type is a `QVariantList` and not `QString`)

Because of the - even low - risk of breaking legacy or external code I would keep it in improved hstore format. Maybe we can change that in future...

Reply:
:+1:  sounds like a good plan. 

Would that also fix this string representation issue https://github.com/qgis/QGIS/issues/30565 ?

Maybe a small utility function that can convert the maps/lists to a decent string representation would be required.

Reply:
I did the string formatting in the widget still. No utility-function, though I like the idea to have one. But since that integration would include the use of this function everywhere else, and we do not have a final decision how we should layout the representing values, I suggest to have it as a separate integration and keep this PR for the fix of the value relation widget. What don't you think @elpaso ?

Reply:
I think that if we don't make a decision we'll hit this kind of issues again and again, and I alto think that having a separate function with full test coverage encourages other developers to not reinvent the (broken) wheel.

Reply:
Okay, thanks. I'll continue on that next week.

Reply:
Hello, let me throw in my 2 cents too ;)

# JSON

I would like to use JSON, but for what I see this will break working with existing data sources. Detecting the format is also no option for what I can see, because when adding a new feature, there is no existing data that could be checked for JSON vs. hstore. So to continue working and keep things stable for our users we *must* continue using hstore for the lifetime of QGIS 3 since I cannot see an automatic way for switching this to JSON transparently. In short, I don't think the risk of hitting legacy issues is low, on the contrary, I would suspect it is almost guaranteed.

Users who want to use something reasonable will be able to setup a data source using native JSON types (which should be working already and is not affected by this pull request)

If we want e.g. shapefile users to use JSON too for this, this should be treated as a - very welcome - "feature request" (new option for the relation widget) and not as a "bug fix".

# Serializing / parsing

If it helps to put hstore serializing/parsing into a public QGIS function, no objections from my side :+1:

Reply:
Thanks.
Yes, I'll make an hstore utility function. 
Any inputs where I should place it?

Reply:
> Yes, I'll make an hstore utility function. Any inputs where I should place it?

https://github.com/qgis/QGIS/blob/master/src/core/qgshstoreutils.h ?

Reply:
> Hello, let me throw in my 2 cents too ;)
> # JSON
> 
> I would like to use JSON, but for what I see this will break working with existing data sources. Detecting the format is also no option for what I can see, because when adding a new feature, there is no existing data that could be checked for JSON vs. hstore. So to continue working and keep things stable for our users we _must_ continue using hstore for the lifetime of QGIS 3 since I cannot see an automatic way for switching this to JSON transparently. In short, I don't think the risk of hitting legacy issues is low, on the contrary, I would suspect it is almost guaranteed.
> 

Well, I partially agree, I think that we have two scenarios here: either we have a new table (or a table with no data in the "map" column) or we have some data in it. If we have some data we could sniff the format and act accordingly.

> Users who want to use something reasonable will be able to setup a data source using native JSON types (which should be working already and is not affected by this pull request)
> 
> If we want e.g. shapefile users to use JSON too for this, this should be treated as a - very welcome - "feature request" (new option for the relation widget) and not as a "bug fix".
> # Serializing / parsing
> 
> If it helps to put hstore serializing/parsing into a public QGIS function, no objections from my side +1
> # Summary
> 
> Let's make sure this pull request fixes things and does not introduce any other regression.

Of course. But I don't get the point: what I'm trying to say here it that this undecided behavior about the internal and the UX (presentation) of complex (map and array values) has been an endless source of issues and apparently when you fix it in one side you break it in another place, it is this attitude to patch it without addressing the real problem than constantly led to this situation (I'm responsible as well, mea culpa).

Additionaly, the UX representation was naive at it best: you cannot just split and join on a comma to get it right.

I do not insist at this point, let's live with it but at least let's find a robust way to represent the values to the user.

Reply:
I have been trying to improve the situation by introducing native types (jsonb etc. for pg, json as data type for gpkg). I think that's fairly stable and I can't see an endless source of issues there.

On the other hand this pull request here is about a regression that plain hstore (pg array) formatted values no longer worked. My only point was that this should be the scope of this PR and that we should eventually be able to work with QGIS 3.10 and 3.4 on the same data without interferences :)

I'm happy if we have the code (possibly with the same deficiencies) working again. If there's a clear path to fix *additional* things, that's good. If it requires more (like cross-checking how exactly postgres handles it etc.) this can easily come in a followup step and should not block this fix of a regression.

Reply:
Regarding the utility function to convert the list in the hstore-like `{1,2,3}` formatted string, I think `qgshstoreutils.h` is not the right place for it. 

Since the naming "hstore" is not very good, because it's not `hstore` - it's how Postgres represents an array. So I would prefer to create a `QgsPgArrayUtils` containing functions to convert a list to this String format and back and try to avoid the naming `hstore` for these formatted strings wherever possible...

Reply:
I agree, the term `hstore` has played a substantial role in muddying the waters.

Reply:
> I would prefer to create a QgsPgArrayUtils

@signedav That's pretty much what I propose [here](https://github.com/qgis/QGIS/pull/9610#issuecomment-510065660) (I try to fix a quite similar issue). I will prefer a non specific Postgres solution. 

We can extract code to parse array from postgres provider (in a new QgsArrayUtils class?). `getNextString` method can already deal with a given separator. It will be easy to improve if some day we want to deal with different format (different brackets, different separator).